### PR TITLE
feat: use interface for events

### PR DESCRIPTION
**Pull request by toddbaert**
Make the providers events member an interface to improve flexibility and reduce possible typing issues.

I'm confident this is non-breaking: I've locally updated the my js-contribs to use this, and everything compiled and tested fine.

### CODE REVIEW

File: packages/client/README.md [hunk 1]
```diff
@@ -298,7 +298,7 @@ class MyProvider implements Provider {
   }
 
   status?: ProviderStatus | undefined;
-  events?: OpenFeatureEventEmitter | undefined;
+  events?: OpenFeatureEventEmitter | undefined; // implement with "new OpenFeatureEventEmitter()", and use "emit()" to emit events
```
Comment (beeme1mr) [inline packages/client/README.md:301]: Should this be `ProviderEventEmitter`?

File: packages/shared/src/events/provider-event-emitter.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  extends ManageLogger<ProviderEventEmitter<E, AdditionalContext>>
+{
+  // here we use E, to restrict the events a provider can manually emit (PROVIDER_CONTEXT_CHANGED is emitted by the SDK)
+  emit(eventType: E, context?: EventContext): void;
+
+  addHandler(eventType: AnyProviderEvent, handler: EventHandler): void;
+
+  removeHandler(eventType: AnyProviderEvent, handler: EventHandler): void;
+
+  removeAllHandlers(eventType?: AnyProviderEvent): void ;
```
Comment (toddbaert) [inline packages/shared/src/events/provider-event-emitter.ts:19]:
```suggestion
  removeAllHandlers(eventType?: AnyProviderEvent): void;
```

File: packages/client/README.md [hunk 3]
```diff
@@ -298,7 +298,7 @@ class MyProvider implements Provider {
   }
 
   status?: ProviderStatus | undefined;
-  events?: OpenFeatureEventEmitter | undefined;
+  events?: ProviderEventEmitter | undefined; // implement with "new OpenFeatureEventEmitter()", and use "emit()" to emit events
```
Comment (jonathannorris) [inline packages/client/README.md:301]:
```suggestion
  
  // implement with "new OpenFeatureEventEmitter()", and use "emit()" to emit events
  events?: ProviderEventEmitter | undefined; 
```

File: packages/server/README.md [hunk 4]
```diff
@@ -302,7 +302,7 @@ class MyProvider implements Provider {
   }
 
   status?: ProviderStatus | undefined;
-  events?: OpenFeatureEventEmitter | undefined;
+  events?: ProviderEventEmitter | undefined; // implement with "new OpenFeatureEventEmitter()", and use "emit()" to emit events
```
Comment (jonathannorris) [inline packages/server/README.md:305]:
```suggestion

  // implement with "new OpenFeatureEventEmitter()", and use "emit()" to emit events
  events?: ProviderEventEmitter | undefined; 
```

**Review by jonathannorris — APPROVED** [review]

File: packages/client/README.md [hunk 5]
```diff
@@ -298,7 +298,9 @@ class MyProvider implements Provider {
   }
 
   status?: ProviderStatus | undefined;
-  events?: OpenFeatureEventEmitter | undefined;
+  
+  // implement with "new OpenFeatureEventEmitter()", and use "emit()" to emit events
+  events?: ProviderEventEmitter | undefined;
```
[the remark this answers is not in the document]
Comment (toddbaert) [reply packages/client/README.md:303]: This is a copy of the interface, nobody will ever write this, it's to get an idea acorss, but I can add the generic, there's no need for the additional context, I think that will only add confusion.

`events?: ProviderEventEmitter<AnyProviderEvent>;` is valid.

Comment (toddbaert) [inline packages/client/README.md:303]:
```suggestion
  events?: ProviderEventEmitter<AnyProviderEvent> | undefined; 
```

File: packages/server/README.md [hunk 6]
```diff
@@ -302,7 +302,9 @@ class MyProvider implements Provider {
   }
 
   status?: ProviderStatus | undefined;
-  events?: OpenFeatureEventEmitter | undefined;
+
+  // implement with "new OpenFeatureEventEmitter()", and use "emit()" to emit events
+  events?: ProviderEventEmitter | undefined;
```
Comment (toddbaert) [inline packages/server/README.md:307]:
```suggestion
  events?: ProviderEventEmitter<AnyProviderEvent> | undefined; 
```

File: packages/client/README.md [hunk 7]
```diff
@@ -298,7 +298,9 @@ class MyProvider implements Provider {
   }
 
   status?: ProviderStatus | undefined;
-  events?: OpenFeatureEventEmitter | undefined;
+  
+  // implement with "new OpenFeatureEventEmitter()", and use "emit()" to emit events
+  events?: ProviderEventEmitter<AnyProviderEvent> | undefined;
```
Comment (beeme1mr) [inline packages/client/README.md:303]: You'll need to import `AnyProviderEvent` 😄

**Review by beeme1mr — APPROVED** [review]

**Review by beeme1mr — APPROVED** [review]

**Review by lukas-reining — APPROVED** [review]
Looks good and also checked it.
Exporting the interface is a good change to allow custom emitters.

**Pull request merged**